File: tdz/src/memory.rs
use crate::{models::*, error::*};
use std::collections::HashMap;
use chrono::Utc;
pub struct MemoryManager {
    pub memories: HashMap<String, Memory>,
    pub memory_tags: HashMap<String, Vec<String>>,
}
impl MemoryManager {
    pub fn new() -> Self {
        Self {
            memories: HashMap::new(),
            memory_tags: HashMap::new(),
        }
    }
    pub async fn create_memory(&mut self, mut memory: Memory) -> Result<String> {
        memory.id = uuid::Uuid::new_v4().to_string();
        memory.created_at = Utc::now();
        memory.updated_at = Utc::now();
        self.memory_tags.insert(memory.id.clone(), memory.tags.clone());
        self.memories.insert(memory.id.clone(), memory.clone());
        Ok(memory.id)
    }
    pub fn get_memory(&self, memory_id: &str) -> Option<&Memory> {
        self.memories.get(memory_id)
    }
    pub fn get_all_memories(&self) -> Vec<&Memory> {
        self.memories.values().collect()
    }
    pub async fn update_memory(
        &mut self,
        memory_id: &str,
        updates: MemoryUpdate,
    ) -> Result<()> {
        if let Some(memory) = self.memories.get_mut(memory_id) {
            if let Some(moment) = updates.moment {
                memory.moment = moment;
            }
            if let Some(meaning) = updates.meaning {
                memory.meaning = meaning;
            }
            if let Some(reason) = updates.reason {
                memory.reason = reason;
            }
            if let Some(importance) = updates.importance {
                memory.importance = importance;
            }
            if let Some(term) = updates.term {
                memory.term = term;
            }
            memory.updated_at = Utc::now();
            if let Some(tags) = updates.tags {
                memory.tags = tags.clone();
                self.memory_tags.insert(memory_id.to_string(), tags);
            }
        } else {
            return Err(TodoziError::ValidationError {
                message: format!("Memory {} not found", memory_id),
            });
        }
        Ok(())
    }
    pub async fn delete_memory(&mut self, memory_id: &str) -> Result<()> {
        if self.memories.remove(memory_id).is_some() {
            self.memory_tags.remove(memory_id);
            Ok(())
        } else {
            Err(TodoziError::ValidationError {
                message: format!("Memory {} not found", memory_id),
            })
        }
    }
    pub fn search_memories(&self, query: &str) -> Vec<&Memory> {
        let query_lower = query.to_lowercase();
        self.memories
            .values()
            .filter(|memory| {
                memory.moment.to_lowercase().contains(&query_lower)
                    || memory.meaning.to_lowercase().contains(&query_lower)
                    || memory.reason.to_lowercase().contains(&query_lower)
                    || memory
                        .tags
                        .iter()
                        .any(|tag| tag.to_lowercase().contains(&query_lower))
            })
            .collect()
    }
    pub fn get_memories_by_importance(
        &self,
        importance: MemoryImportance,
    ) -> Vec<&Memory> {
        self.memories.values().filter(|memory| memory.importance == importance).collect()
    }
    pub fn get_memories_by_term(&self, term: MemoryTerm) -> Vec<&Memory> {
        self.memories.values().filter(|memory| memory.term == term).collect()
    }
    pub fn get_memories_by_tag(&self, tag: &str) -> Vec<&Memory> {
        let tag_lower = tag.to_lowercase();
        self.memories
            .values()
            .filter(|memory| memory.tags.iter().any(|t| t.to_lowercase() == tag_lower))
            .collect()
    }
    pub fn get_recent_memories(&self, limit: usize) -> Vec<&Memory> {
        let mut memories: Vec<&Memory> = self.memories.values().collect();
        memories.sort_by(|a, b| b.created_at.cmp(&a.created_at));
        memories.into_iter().take(limit).collect()
    }
    pub fn get_critical_memories(&self) -> Vec<&Memory> {
        self.memories
            .values()
            .filter(|memory| {
                memory.importance == MemoryImportance::High
                    || memory.importance == MemoryImportance::Critical
            })
            .collect()
    }
    pub fn get_short_term_memories(&self) -> Vec<&Memory> {
        self.memories
            .values()
            .filter(|memory| memory.term == MemoryTerm::Short)
            .collect()
    }
    pub fn get_long_term_memories(&self) -> Vec<&Memory> {
        self.memories.values().filter(|memory| memory.term == MemoryTerm::Long).collect()
    }
    pub fn get_memories_by_type(&self, memory_type: &MemoryType) -> Vec<&Memory> {
        self.memories
            .values()
            .filter(|memory| &memory.memory_type == memory_type)
            .collect()
    }
    pub fn get_secret_memories(&self) -> Vec<&Memory> {
        self.get_memories_by_type(&MemoryType::Secret)
    }
    pub fn get_human_memories(&self) -> Vec<&Memory> {
        self.get_memories_by_type(&MemoryType::Human)
    }
    pub fn get_emotional_memories(&self, emotion: &str) -> Vec<&Memory> {
        self.memories
            .values()
            .filter(|memory| {
                if let MemoryType::Emotional(mem_emotion) = &memory.memory_type {
                    mem_emotion == emotion
                } else {
                    false
                }
            })
            .collect()
    }
    pub fn get_all_tags(&self) -> Vec<String> {
        let mut all_tags = std::collections::HashSet::new();
        for tags in self.memory_tags.values() {
            for tag in tags {
                all_tags.insert(tag.clone());
            }
        }
        all_tags.into_iter().collect()
    }
    pub fn get_tag_statistics(&self) -> HashMap<String, usize> {
        let mut stats = HashMap::new();
        for tags in self.memory_tags.values() {
            for tag in tags {
                *stats.entry(tag.clone()).or_insert(0) += 1;
            }
        }
        stats
    }
    pub fn get_memory_statistics(&self) -> MemoryStatistics {
        let total_memories = self.memories.len();
        let short_term = self.get_short_term_memories().len();
        let long_term = self.get_long_term_memories().len();
        let critical = self.get_critical_memories().len();
        let unique_tags = self.get_all_tags().len();
        let secret = self.get_secret_memories().len();
        let human = self.get_human_memories().len();
        let standard = self.get_memories_by_type(&MemoryType::Standard).len();
        let mut emotional = 0;
        let emotions = vec![
            "happy", "sad", "angry", "fearful", "surprised", "disgusted", "excited",
            "anxious", "confident", "frustrated", "motivated", "overwhelmed", "curious",
            "satisfied", "disappointed", "grateful", "proud", "ashamed", "hopeful",
            "resigned"
        ];
        for emotion in emotions {
            emotional += self.get_emotional_memories(emotion).len();
        }
        MemoryStatistics {
            total_memories,
            short_term_memories: short_term,
            long_term_memories: long_term,
            critical_memories: critical,
            unique_tags,
            secret_memories: secret,
            human_memories: human,
            emotional_memories: emotional,
            standard_memories: standard,
        }
    }
}
#[derive(Debug, Clone)]
pub struct MemoryUpdate {
    pub moment: Option<String>,
    pub meaning: Option<String>,
    pub reason: Option<String>,
    pub importance: Option<MemoryImportance>,
    pub term: Option<MemoryTerm>,
    pub tags: Option<Vec<String>>,
}
impl MemoryUpdate {
    pub fn new() -> Self {
        Self {
            moment: None,
            meaning: None,
            reason: None,
            importance: None,
            term: None,
            tags: None,
        }
    }
    pub fn moment(mut self, moment: String) -> Self {
        self.moment = Some(moment);
        self
    }
    pub fn meaning(mut self, meaning: String) -> Self {
        self.meaning = Some(meaning);
        self
    }
    pub fn reason(mut self, reason: String) -> Self {
        self.reason = Some(reason);
        self
    }
    pub fn importance(mut self, importance: MemoryImportance) -> Self {
        self.importance = Some(importance);
        self
    }
    pub fn term(mut self, term: MemoryTerm) -> Self {
        self.term = Some(term);
        self
    }
    pub fn tags(mut self, tags: Vec<String>) -> Self {
        self.tags = Some(tags);
        self
    }
}
#[derive(Debug, Clone)]
pub struct MemoryStatistics {
    pub total_memories: usize,
    pub short_term_memories: usize,
    pub long_term_memories: usize,
    pub critical_memories: usize,
    pub unique_tags: usize,
    pub secret_memories: usize,
    pub human_memories: usize,
    pub emotional_memories: usize,
    pub standard_memories: usize,
}
impl MemoryStatistics {
    pub fn short_term_percentage(&self) -> f64 {
        if self.total_memories == 0 {
            0.0
        } else {
            (self.short_term_memories as f64 / self.total_memories as f64) * 100.0
        }
    }
    pub fn long_term_percentage(&self) -> f64 {
        if self.total_memories == 0 {
            0.0
        } else {
            (self.long_term_memories as f64 / self.total_memories as f64) * 100.0
        }
    }
    pub fn critical_percentage(&self) -> f64 {
        if self.total_memories == 0 {
            0.0
        } else {
            (self.critical_memories as f64 / self.total_memories as f64) * 100.0
        }
    }
}
pub fn parse_memory_format(memory_text: &str, user_id: &str) -> Result<Memory> {
    let start_tag = "<memory>";
    let end_tag = "</memory>";
    let start = memory_text
        .find(start_tag)
        .ok_or_else(|| TodoziError::ValidationError {
            message: "Missing <memory> start tag".to_string(),
        })?;
    let end = memory_text
        .find(end_tag)
        .ok_or_else(|| TodoziError::ValidationError {
            message: "Missing </memory> end tag".to_string(),
        })?;
    let content = &memory_text[start + start_tag.len()..end];
    let parts: Vec<&str> = content.split(';').map(|s| s.trim()).collect();
    if parts.len() < 6 {
        return Err(TodoziError::ValidationError {
            message: "Invalid memory format: need at least 6 parts (type; moment; meaning; reason; importance; term)"
                .to_string(),
        });
    }
    let memory_type_str = parts[0];
    let emotion_list = vec![
        "happy", "sad", "angry", "fearful", "surprised", "disgusted", "excited",
        "anxious", "confident", "frustrated", "motivated", "overwhelmed", "curious",
        "satisfied", "disappointed", "grateful", "proud", "ashamed", "hopeful",
        "resigned",
    ];
    let memory_type = if emotion_list.contains(&memory_type_str) {
        MemoryType::Emotional(memory_type_str.to_string())
    } else {
        match memory_type_str {
            "standard" => MemoryType::Standard,
            "secret" => MemoryType::Secret,
            "human" => MemoryType::Human,
            "short" => MemoryType::Short,
            "long" => MemoryType::Long,
            _ => MemoryType::Standard,
        }
    };
    let tags = if parts.len() > 6 && !parts[6].is_empty() {
        parts[6].split(',').map(|s| s.trim().to_string()).collect()
    } else {
        Vec::new()
    };
    Ok(Memory {
        id: uuid::Uuid::new_v4().to_string(),
        user_id: user_id.to_string(),
        project_id: None,
        status: ItemStatus::Active,
        moment: parts[1].to_string(),
        meaning: parts[2].to_string(),
        reason: parts[3].to_string(),
        importance: parts[4]
            .parse::<MemoryImportance>()
            .map_err(|_| TodoziError::ValidationError {
                message: "Invalid memory importance".to_string(),
            })?,
        term: parts[5]
            .parse::<MemoryTerm>()
            .map_err(|_| TodoziError::ValidationError {
                message: "Invalid memory term".to_string(),
            })?,
        memory_type,
        tags,
        created_at: Utc::now(),
        updated_at: Utc::now(),
    })
}
#[cfg(test)]
mod tests {
    use super::*;
    #[test]
    fn test_memory_manager_creation() {
        let manager = MemoryManager::new();
        assert_eq!(manager.memories.len(), 0);
        assert_eq!(manager.memory_tags.len(), 0);
    }
    #[test]
    fn test_memory_update_builder() {
        let update = MemoryUpdate::new()
            .moment("New moment".to_string())
            .meaning("New meaning".to_string())
            .importance(MemoryImportance::High);
        assert_eq!(update.moment, Some("New moment".to_string()));
        assert_eq!(update.meaning, Some("New meaning".to_string()));
        assert_eq!(update.importance, Some(MemoryImportance::High));
    }
    #[test]
    fn test_memory_statistics_percentages() {
        let stats = MemoryStatistics {
            total_memories: 10,
            short_term_memories: 6,
            long_term_memories: 4,
            critical_memories: 2,
            unique_tags: 8,
            secret_memories: 1,
            human_memories: 2,
            emotional_memories: 3,
            standard_memories: 4,
        };
        assert_eq!(stats.short_term_percentage(), 60.0);
        assert_eq!(stats.long_term_percentage(), 40.0);
        assert_eq!(stats.critical_percentage(), 20.0);
        let empty_stats = MemoryStatistics {
            total_memories: 0,
            short_term_memories: 0,
            long_term_memories: 0,
            critical_memories: 0,
            unique_tags: 0,
            secret_memories: 0,
            human_memories: 0,
            emotional_memories: 0,
            standard_memories: 0,
        };
        assert_eq!(empty_stats.short_term_percentage(), 0.0);
        assert_eq!(empty_stats.long_term_percentage(), 0.0);
        assert_eq!(empty_stats.critical_percentage(), 0.0);
    }
    #[test]
    fn test_parse_memory_format() {
        let memory_text = "<memory>standard; 2025-01-13 10:30 AM; Client prefers iterative development; Affects testing cycle; high; long; client,development,iterative</memory>";
        let memory = parse_memory_format(memory_text, "user_123").unwrap();
        assert_eq!(memory.moment, "2025-01-13 10:30 AM");
        assert_eq!(memory.meaning, "Client prefers iterative development");
        assert_eq!(memory.reason, "Affects testing cycle");
        assert_eq!(memory.importance, MemoryImportance::High);
        assert_eq!(memory.term, MemoryTerm::Long);
        assert_eq!(memory.memory_type, MemoryType::Standard);
        assert_eq!(memory.tags, vec!["client", "development", "iterative"]);
    }
}

File: tdz/py/memory.py
"""Memory management for Todozi."""

from typing import Dict, List, Optional
from datetime import datetime
import uuid as uuid_lib

from .models import Memory, MemoryImportance, MemoryTerm, MemoryType, ItemStatus
from .error import ValidationError


class MemoryManager:
    """Manages memories."""

    def __init__(self):
        self.memories: Dict[str, Memory] = {}
        self.memory_tags: Dict[str, List[str]] = {}

    async def create_memory(self, memory: Memory) -> str:
        """Create a new memory."""
        memory.id = str(uuid_lib.uuid4())
        memory.created_at = datetime.utcnow()
        memory.updated_at = datetime.utcnow()
        self.memory_tags[memory.id] = memory.tags.copy()
        self.memories[memory.id] = memory
        return memory.id

    def get_memory(self, memory_id: str) -> Optional[Memory]:
        """Get a memory by ID."""
        return self.memories.get(memory_id)

    def get_all_memories(self) -> List[Memory]:
        """Get all memories."""
        return list(self.memories.values())

    async def update_memory(self, memory_id: str, updates: 'MemoryUpdate'):
        """Update a memory."""
        memory = self.memories.get(memory_id)
        if not memory:
            raise ValidationError(f"Memory {memory_id} not found")

        if updates.moment is not None:
            memory.moment = updates.moment
        if updates.meaning is not None:
            memory.meaning = updates.meaning
        if updates.reason is not None:
            memory.reason = updates.reason
        if updates.importance is not None:
            memory.importance = updates.importance
        if updates.term is not None:
            memory.term = updates.term
        if updates.tags is not None:
            memory.tags = updates.tags.copy()
            self.memory_tags[memory_id] = updates.tags

        memory.updated_at = datetime.utcnow()

    async def delete_memory(self, memory_id: str):
        """Delete a memory."""
        if memory_id not in self.memories:
            raise ValidationError(f"Memory {memory_id} not found")
        del self.memories[memory_id]
        self.memory_tags.pop(memory_id, None)

    def search_memories(self, query: str) -> List[Memory]:
        """Search for memories."""
        query_lower = query.lower()
        return [
            m for m in self.memories.values()
            if query_lower in m.moment.lower()
            or query_lower in m.meaning.lower()
            or query_lower in m.reason.lower()
            or any(query_lower in tag.lower() for tag in m.tags)
        ]

    def get_memories_by_importance(self, importance: MemoryImportance) -> List[Memory]:
        """Get memories by importance level."""
        return [m for m in self.memories.values() if m.importance == importance]

    def get_memories_by_term(self, term: MemoryTerm) -> List[Memory]:
        """Get memories by term."""
        return [m for m in self.memories.values() if m.term == term]

    def get_memories_by_tag(self, tag: str) -> List[Memory]:
        """Get memories by tag."""
        tag_lower = tag.lower()
        return [
            m for m in self.memories.values()
            if any(t.lower() == tag_lower for t in m.tags)
        ]

    def get_recent_memories(self, limit: int) -> List[Memory]:
        """Get recent memories."""
        memories = sorted(self.memories.values(), key=lambda x: x.created_at, reverse=True)
        return memories[:limit]

    def get_critical_memories(self) -> List[Memory]:
        """Get critical memories."""
        return [
            m for m in self.memories.values()
            if m.importance in [MemoryImportance.HIGH, MemoryImportance.CRITICAL]
        ]

    def get_short_term_memories(self) -> List[Memory]:
        """Get short-term memories."""
        return [m for m in self.memories.values() if m.term == MemoryTerm.SHORT]

    def get_long_term_memories(self) -> List[Memory]:
        """Get long-term memories."""
        return [m for m in self.memories.values() if m.term == MemoryTerm.LONG]

    def get_memories_by_type(self, memory_type: MemoryType) -> List[Memory]:
        """Get memories by type."""
        return [m for m in self.memories.values() if m.memory_type == memory_type]

    def get_secret_memories(self) -> List[Memory]:
        """Get secret memories."""
        return self.get_memories_by_type(MemoryType.SECRET)

    def get_human_memories(self) -> List[Memory]:
        """Get human memories."""
        return self.get_memories_by_type(MemoryType.HUMAN)

    def get_all_tags(self) -> List[str]:
        """Get all unique tags."""
        all_tags = set()
        for tags in self.memory_tags.values():
            all_tags.update(tags)
        return list(all_tags)

    def get_tag_statistics(self) -> Dict[str, int]:
        """Get tag usage statistics."""
        stats = {}
        for tags in self.memory_tags.values():
            for tag in tags:
                stats[tag] = stats.get(tag, 0) + 1
        return stats

    def get_memory_statistics(self) -> 'MemoryStatistics':
        """Get memory statistics."""
        return MemoryStatistics(
            total_memories=len(self.memories),
            short_term_memories=len(self.get_short_term_memories()),
            long_term_memories=len(self.get_long_term_memories()),
            critical_memories=len(self.get_critical_memories()),
            unique_tags=len(self.get_all_tags()),
            secret_memories=len(self.get_secret_memories()),
            human_memories=len(self.get_human_memories()),
            emotional_memories=0,  # Count emotional memories
            standard_memories=len(self.get_memories_by_type(MemoryType.STANDARD))
        )


class MemoryUpdate:
    """Memory update structure."""
    def __init__(self):
        self.moment: Optional[str] = None
        self.meaning: Optional[str] = None
        self.reason: Optional[str] = None
        self.importance: Optional[MemoryImportance] = None
        self.term: Optional[MemoryTerm] = None
        self.tags: Optional[List[str]] = None


class MemoryStatistics:
    """Memory statistics."""
    def __init__(self, total_memories: int, short_term_memories: int,
                 long_term_memories: int, critical_memories: int,
                 unique_tags: int, secret_memories: int, human_memories: int,
                 emotional_memories: int, standard_memories: int):
        self.total_memories = total_memories
        self.short_term_memories = short_term_memories
        self.long_term_memories = long_term_memories
        self.critical_memories = critical_memories
        self.unique_tags = unique_tags
        self.secret_memories = secret_memories
        self.human_memories = human_memories
        self.emotional_memories = emotional_memories
        self.standard_memories = standard_memories

    def short_term_percentage(self) -> float:
        """Get percentage of short-term memories."""
        return (self.short_term_memories / self.total_memories * 100.0) if self.total_memories > 0 else 0.0

    def long_term_percentage(self) -> float:
        """Get percentage of long-term memories."""
        return (self.long_term_memories / self.total_memories * 100.0) if self.total_memories > 0 else 0.0

    def critical_percentage(self) -> float:
        """Get percentage of critical memories."""
        return (self.critical_memories / self.total_memories * 100.0) if self.total_memories > 0 else 0.0


def parse_memory_format(memory_text: str, user_id: str) -> Memory:
    """Parse memory from XML-like format."""
    start_tag = "<memory>"
    end_tag = "</memory>"

    start = memory_text.find(start_tag)
    if start == -1:
        raise ValidationError("Missing <memory> start tag")

    end = memory_text.find(end_tag)
    if end == -1:
        raise ValidationError("Missing </memory> end tag")

    content = memory_text[start + len(start_tag):end]
    parts = [s.strip() for s in content.split(';')]

    if len(parts) < 6:
        raise ValidationError(
            "Invalid memory format: need at least 6 parts (type; moment; meaning; reason; importance; term)"
        )

    # Parse memory type
    memory_type_str = parts[0]
    if memory_type_str in ["standard", "secret", "human", "short", "long"]:
        memory_type = MemoryType(memory_type_str)
    else:
        # Assume it's an emotional type
        memory_type = MemoryType.emotional(memory_type_str)

    # Parse tags
    tags = []
    if len(parts) > 6 and parts[6]:
        tags = [s.strip() for s in parts[6].split(',')]

    return Memory(
        id=str(uuid_lib.uuid4()),
        user_id=user_id,
        project_id=None,
        status=ItemStatus.ACTIVE,
        moment=parts[1],
        meaning=parts[2],
        reason=parts[3],
        importance=MemoryImportance(parts[4].lower()),
        term=MemoryTerm(parts[5].lower()),
        memory_type=memory_type,
        tags=tags,
        created_at=datetime.utcnow(),
        updated_at=datetime.utcnow()
    )
